Make python test suites run in parallel

jtattermusch opened the pull request:
-- allow filtering test cases using a prefix in _runner.py
-- added tests.json file that contains list of all existing python test suites (the list is used by run_tests.py to determine which suites to run). To make sure that the tests.json file is always up-to-date, I added a "sanity" tests that discovers all existings tests using test loader and checks that tests.json is up-to-date.
-- gcov still runs all the tests serially and should have the same behavior as before

The work is not 100% finished yet, but would be worth reviewing.

jtattermusch commented:
@soltanmm  please also take a look.

jtattermusch commented:
@ctiller  @nicolasnoble  FYI

soltanmm commented on tools/run_tests/run_python.sh:
```diff
@@ -42,7 +42,12 @@ export LDFLAGS="-L$ROOT/libs/$CONFIG"
 export GRPC_PYTHON_ENABLE_CYTHON_TRACING=1
 
-tox
+if [ "$CONFIG" = "gcov" ]
```
I'm confused. What's this branch for?

jtattermusch replied:
for code coverage we use the commands from tox.ini to run the tests and generate the coverage report.
For a "regular test run" (=opt or dbg), we are running tests in parallel and tox cannot handle that. So we make sure everything is built in advance and just run the test here.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/_sanity/_sanity_test.py:
```diff
@@ -0,0 +1,51 @@
+    """Autodiscovers all test suites and checks that tests.json is up to date"""
+    loader = _loader.Loader()
+    loader.loadTestsFromNames(['tests'])
+    test_suite_names = [c.id().rsplit('.', 1)[0]
```
Please avoid single-letter identifiers.

The values assigned to `c` are `unittest.TestCase` subclasses, right? How about `test_case_class` for a name? You can win some horizontal space by line-wrapping immediately after `[`:
```
    test_suite_names = [
        test_case_class.id().rsplit('.', 1)[0]
        for test_case_class in _loader.iterate_suite_cases(loader.suite)]
```
.

jtattermusch replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/_sanity/_sanity_test.py:
```diff
@@ -0,0 +1,51 @@
+                        for c in _loader.iterate_suite_cases(loader.suite)]
+    test_suite_names = sorted(set(test_suite_names))
+
+    with open('src/python/grpcio/tests/tests.json') as f:
```
`tests_json_file` rather than `f`, please.

jtattermusch replied:
done

nathanielmanistaatgoogle commented on tools/run_tests/run_tests.py:
```diff
@@ -353,15 +353,27 @@ def configure(self, config, args):
 
   def test_specs(self):
+    # load list of known test suites
+    with open('src/python/grpcio/tests/tests.json') as f:
```
`tests_json_file` here too?

jtattermusch replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/_sanity/_sanity_test.py:
```diff
@@ -0,0 +1,51 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+from tests import _loader
```
Standard-library imports come first, then local imports. Also you should import any underscore-prefixed attribute of another code element.

I think these imports should be
```
import json
import unittest

import tests
```
and then below where you currently have `_loader.Loader()` you would just have `tests.Loader()`.

jtattermusch replied:
done

nathanielmanistaatgoogle commented:
Impressive; I like what you've done here.

jtattermusch commented:
Ok, I addressed all the comments and squashed the commits. Let's see how the test results look like.
Also, thanks for a little lesson of good python style.

nathanielmanistaatgoogle commented:
Commit content looks good, commit itself looks good, Python is green, sanity is [`test/cpp/interop/reconnect_interop_client.cc modified 2016 but copyright only extends to 2015`](https://en.wikipedia.org/wiki/Somebody_else's_problem). I'm satisfied with this but @soltanmm will also review and approve before merge.

soltanmm commented:
:thumbsup:

I'm fine with this for now.
